eth/utils/bls.py

```diff
@@ -20,124 +24,200 @@
+        )
     x, y = normalize(pt)
     return (
-        int(x.coeffs[0] + 2**255 * (y.coeffs[0] % 2)),
+        int(x.coeffs[0] + 2**383 * (y.coeffs[0] % 2)),
         int(x.coeffs[1])
     )
 
 
 def decompress_G2(p: bytes) -> Tuple[FQP, FQP, FQP]:
-    x1 = p[0] % 2**255
-    y1_mod_2 = p[0] // 2**255
+    x1 = p[0] % 2**383
+    y1_mod_2 = p[0] // 2**383
     x2 = p[1]
     x = FQ2([x1, x2])
     if x == FQ2([0, 0]):
         return FQ2([1, 0]), FQ2([1, 0]), FQ2([0, 0])
-    y = sqrt_fq2(x**3 + b2)
+    y = modular_squareroot(x**3 + b2)
     if y.coeffs[0] % 2 != y1_mod_2:
         y = FQ2((y * -1).coeffs)
-    assert is_on_curve((x, y, FQ2([1, 0])), b2)
+    if not is_on_curve((x, y, FQ2([1, 0])), b2):
+        raise ValueError(
+            "The given point is not on the twisted curve over FQ**2"
+        )
     return x, y, FQ2([1, 0])
 
 
-def sign(m: bytes, k: int) -> Tuple[int, int]:
-    return compress_G2(multiply(hash_to_G2(m), k))
+#
+# APIs
+#
+def sign(message: bytes,
+         privkey: int,
+         domain: int) -> Tuple[int, int]:
+    return compress_G2(
+        multiply(
+            hash_to_G2(message, domain),
+            privkey
+        )
+    )
 
 
 def privtopub(k: int) -> int:
     return compress_G1(multiply(G1, k))
 
 
-def verify(m: bytes, pub: int, sig: bytes) -> bool:
+def verify(message: bytes, pubkey: int, signature: bytes, domain: int) -> bool:
     final_exponentiation = final_exponentiate(
-        pairing(FQP_point_to_FQ2_point(decompress_G2(sig)), G1, False) *
-        pairing(FQP_point_to_FQ2_point(hash_to_G2(m)), neg(decompress_G1(pub)), False)
+        pairing(FQP_point_to_FQ2_point(decompress_G2(signature)), G1, False) *
+        pairing(
+            FQP_point_to_FQ2_point(hash_to_G2(message, domain)),
+            neg(decompress_G1(pubkey)),
+            False
+        )
     )
     return final_exponentiation == FQ12.one()
 
 
-def aggregate_sigs(sigs: Iterable[bytes]) -> Tuple[int, int]:
+def aggregate_signatures(signatures: Sequence[bytes]) -> Tuple[int, int]:
     o = Z2
-    for s in sigs:
+    for s in signatures:
         o = FQP_point_to_FQ2_point(add(o, decompress_G2(s)))
     return compress_G2(o)
 
 
-def aggregate_pubs(pubs: Iterable[int]) -> int:
+def aggregate_pubkeys(pubkeys: Sequence[int]) -> int:
     o = Z1
-    for p in pubs:
+    for p in pubkeys:
         o = add(o, decompress_G1(p))
     return compress_G1(o)
+
+
+def verify_multiple(pubkeys: Sequence[int],
+                    messages: Sequence[bytes],
+                    signature: bytes,
+                    domain: int) -> bool:
+    len_msgs = len(messages)
+
+    if len(pubkeys) != len_msgs:
+        raise ValidationError(
+            "len(pubkeys) (%s) should be equal to len(messages) (%s)" % (
+                len(pubkeys), len_msgs
+            )
+        )
+
+    o = FQ12([1] + [0] * 11)
+    for m_pubs in set(messages):
+        # aggregate the pubs
+        group_pub = Z1
```

Comment:
I think I was a bit unclear in the last comment. I meant something like the following which utilizes `aggregate_pubs` rather than directly calling `add`

```python
for message in set(messages):
    pubkeys_for_message = [
        pubkey for i, pubkey in enumerate(pubkeys)
        if messages[i] == message
    ]
    group_pub = aggregate_pubkeys(pubkeys_for_message)
```

Reply:
Oh I see! I think I've tried to do that before, if we want to utilize `aggregate_pubkeys`, we will need:
1. `len(messages)` times extra `compress_G1()` calling (inside `aggregate_pubkeys`)
2. `len(messages)` times extra `decompress_G1()` calling for the second argument of `pairing`.

Alternatively, we might refactor `aggregate_signatures(signatures: Sequence[bytes]) -> Tuple[int, int]` to:
```python
def aggregate_pubkeys(pubkeys: Sequence[int]) -> int:
    return compress_G1(_aggregate_pubkeys(pubkeys))

def _aggregate_pubkeys(pubkeys: Sequence[int]) -> Tuple[FQ, FQ, FQ]:
    o = Z1
    for p in pubkeys:
        o = add(o, decompress_G1(p))
    return o
```

And then make `verify_multiple` call `_aggregate_pubkeys()`.


What do you think of it?

Reply:
oh crap, you're right. 

I'm okay either way. It's not a huge gain in code reuse and this code won't change much once in place.

Reply:
Do whichever you want. I'm going to use the updated work in #1631 in the morning so merge whenever you're ready.